crates/transport/src/layers/fallback.rs

```diff
@@ -0,0 +1,447 @@
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.partial_cmp(other).unwrap_or(std::cmp::Ordering::Equal)
+    }
+}
+
+/// The fallback provider consumes multiple transports and uses a
+/// ranking strategy to provide responses optimizing for the best
+/// latency and availability.
+#[derive(Debug, Clone)]
+pub struct FallbackService<S> {
+    transports: Arc<Mutex<BinaryHeap<ScoredTransport<S>>>>,
+    active_transport_count: usize,
+    log_transport_rankings: bool,
+}
+
+impl<S> FallbackService<S> {
+    /// Create a new fallback provider from a list of transports.
+    ///
+    /// - The `active_transport_count` parameter controls how many transports are used for requests
+    ///   at any one time.
+    /// - The `log_transport_rankings` parameter controls whether the current transport rankings are
+    ///   output as traces. In order to see them, make sure to enable the `alloy_transport` target
+    ///   (i.e. `RUST_LOG=alloy_transport=debug`).
+    pub fn new(
+        transports: Vec<S>,
+        active_transport_count: usize,
+        log_transport_rankings: bool,
+    ) -> Self {
+        let scored_transports = transports
+            .into_iter()
+            .enumerate()
+            .map(|(id, transport)| ScoredTransport::new(id, transport))
+            .collect::<Vec<_>>();
+
+        Self {
+            transports: Arc::new(Mutex::new(BinaryHeap::from(scored_transports))),
+            active_transport_count,
+            log_transport_rankings,
+        }
+    }
+
+    /// Get the number of transports
+    pub fn transport_count(&self) -> usize {
+        let transports = self.transports.lock().expect("Lock poisoned");
+        transports.len()
+    }
+
+    /// Log the current ranking of transports
+    pub fn log_transport_rankings(&self)
+    where
+        S: std::fmt::Debug,
+    {
+        let transports = self.transports.lock().expect("Lock poisoned");
+        let mut sorted_transports: Vec<_> = transports.iter().collect();
+        sorted_transports.sort_by(|a, b| b.cmp(a)); // Sort by score (descending)
+
+        debug!("Current transport rankings:");
+        for (idx, transport) in sorted_transports.iter().enumerate() {
+            debug!("  #{}: Transport[{}] - {}", idx + 1, transport.id, transport.metrics_summary());
+        }
+    }
+}
+
+impl<S> FallbackService<S>
+where
+    S: Service<RequestPacket, Future = TransportFut<'static>, Error = TransportError>
+        + Send
+        + Clone
+        + std::fmt::Debug
+        + 'static,
+{
+    async fn make_request(&self, req: RequestPacket) -> Result<ResponsePacket, TransportError> {
+        let mut top_transports = Vec::new();
+        {
+            let mut transports = self.transports.lock().expect("Lock poisoned");
+            for _ in 0..self.active_transport_count.min(transports.len()) {
+                if let Some(transport) = transports.pop() {
+                    top_transports.push(transport);
```

Comment:
not exactly sure I fully understood this but it seems like this pops the transports and on success it adds them back?

what happens if there are no available transports?

Reply:
Yes, I've updated this to use a `Vec` instead of a `BinaryHeap` and re-sort it whenever a score changes. 
This way I don't need to take the transports from the list on each request.